TODO: Parse ``listen_addrs`` to set transport in ``build_swarm``

lla-dane opened the pull request:
## Discription 
Extended the [new_swarm](https://github.com/libp2p/py-libp2p/blob/5b40e2551dd70f8a8e3fd965e04aacc11ec3b248/libp2p/__init__.py#L61C1-L67C1) function's parameters and logic  to include optional `listen_addrs`, which will be parsed to set the transport protocol for the swarm.

lla-dane commented:
@seetadev @acul71

lla-dane commented on libp2p/__init__.py:
```diff
@@ -78,8 +84,18 @@ def new_swarm(
+    if listen_addrs is None:
+        transport = TCP()
+    else:
+        addr = listen_addrs[0]
```
Here assumed that all the `listen_addrs` will have the same transport protocol

lla-dane commented on examples/ping/ping.py:
```diff
@@ -57,7 +57,7 @@ async def send_ping(stream: INetStream) -> None:
     localhost_ip = "127.0.0.1"
     listen_addr = multiaddr.Multiaddr(f"/ip4/0.0.0.0/tcp/{port}")
-    host = new_host()
+    host = new_host(listen_addrs=[listen_addr])
```
The `ping` example works fine with this modification.

pacrob commented:
Thanks, @lla-dane! Overall looks good, just a couple things:

 - You'll want to update the docstrings with the new arg.
 - Add some tests to show that defaults and exceptions are behaving as you expect.
 - And add a newsfragment.

The failing tests don't seem to be related.

acul71 commented:
> The failing tests don't seem to be related.

Yes the `FAILED tests/core/network/test_notify.py::test_notify` has been fixed in other two  PRs that aren't merged yet

acul71 commented:
> Thanks, @lla-dane! Overall looks good, just a couple things:
> 
> * You'll want to update the docstrings with the new arg.
> * Add some tests to show that defaults and exceptions are behaving as you expect.
> * And add a newsfragment.

Hi @lla-dane if you need help how to do that just ping me here or on discord, don't be shy. Ciao Luca

lla-dane commented:
> * You'll want to update the docstrings with the new arg.
> * Add some tests to show that defaults and exceptions are behaving as you expect.
> * And add a newsfragment.

Added the tests and newfragments and updated the docstrings. @acul71 @pacrob.
Only left the options for `quic` and `tcp`, because the other cases were raising error while parsing through `multiaddr.Multiaddr()`

pacrob commented on libp2p/__init__.py:
```diff
@@ -77,15 +83,22 @@ def new_swarm(
+        if addr.__contains__("tcp"):
+            transport = TCP()
+        elif addr.__contains__("quic"):
+            raise ValueError("QUIC not yet supported")
```
We'll want an `else` at the bottom here to catch the case where it's not `None`, `tcp`, or `quic`.

pacrob replied:
```suggestion
            raise ValueError("QUIC not yet supported")
        else:
            raise ValueError(f"listen_addrs contains unknown transport: {listen_addrs}")
```

lla-dane commented:
Do I need to squash all the commits into one ?

pacrob commented:
> Do I need to squash all the commits into one ?

Yes please!

lla-dane commented:
> Yes please!

Done

acul71 commented:
> I don't see why this test is failing after squashing the commits
`FAILED tests/core/network/test_notify.py::test_notify`
Because there are some race conditions sometimes it fails somtimes it does not
https://github.com/libp2p/py-libp2p/pull/614#issuecomment-2890383837

lla-dane commented:
Rebased with the last commit, the failing test was updated, it should be running successfully this time.
